editor/osm_feature_matcher.cpp

```diff
@@ -0,0 +1,145 @@
+#include "base/logging.hpp"
+
+#include "editor/osm_feature_matcher.hpp"
+
+namespace osm
+{
+double constexpr kPointDiffEps = MercatorBounds::GetCellID2PointAbsEpsilon();
+
+bool LatLonEqual(ms::LatLon const & a, ms::LatLon const & b)
+{
+  return a.EqualDxDy(b, kPointDiffEps);
+}
+
+double ScoreLatLon(XMLFeature const & xmlFt, ms::LatLon const & latLon)
+{
+  auto const a = MercatorBounds::FromLatLon(xmlFt.GetCenter());
+  auto const b = MercatorBounds::FromLatLon(latLon);
+  return 1.0 - (a.Length(b) / kPointDiffEps);
+}
+
+double ScoreGeometry(pugi::xml_document const & osmResponse, pugi::xml_node const & way,
+                     set<m2::PointD> geometry)
+{
+  int matched = 0;
+  int total = 0;
+  // TODO(mgsergio): optimize using sorting and scaning eps-squares.
+  // The idea is to build squares with side = eps on points from the first set.
+  // Sort them by y and x in y groups.
+  // Then pass points from the second set through constructed struct and register events
+  // like square started, square ended, point encounted.
+
+  for (auto const xNodeRef : way.select_nodes("nd/@ref"))
+  {
+    ++total;
+    string const nodeRef = xNodeRef.attribute().value();
+    auto const node = osmResponse.select_node(("osm/node[@id='" + nodeRef + "']").data()).node();
+    if (!node)
+    {
+      LOG(LDEBUG, ("OSM response have ref", nodeRef,
+                   "but have no node with such id.", osmResponse));
+      continue;  // TODO: or break and return -1000?
+    }
+
+    XMLFeature xmlFt(node);
+    for (auto pointIt = begin(geometry); pointIt != end(geometry); ++pointIt)
```

Comment:
In general, it's not performance-critical: this code is rarely executed in background, when user has made any map edits and finished working with the app. Code looks cleaner in this way.

Btw I personally prefer geometry.begin() instead of begin(geometry).